torantie's solution for Project

torantie opened the pull request:
Unfortunately could not find the error in the Mean value weights calculation.

Commit:
Convert to halfedge

torantie committed:
Update README.md

Commit:
With Comments. Mean and k2 not working

Commit:
Refactoring

Commit:
Refactoring

Commit:
Refactoring

Commit:
Refactoring

Commit:
Refactoring

Commit:
Change Angle Defect

Commit:
Revert "Convert to halfedge"

This reverts commit 7a942c3834504a62330f1bc1ae9c470e7e9b5eaf.

torantie committed:
Merge remote-tracking branch 'upstream/master'

torantie committed:
Merge remote-tracking branch 'upstream/master'

upstream merge

torantie committed:
cleanup

torantie committed:
start

torantie committed:
intermediate state

torantie committed:
intermediate state

torantie committed:
intermediate state

torantie committed:
intermediate state

torantie committed:
intermediate state. works

torantie committed:
intermediate state.

torantie committed:
intermediate state. works

torantie committed:
intermediate state. works. Main bug

torantie committed:
intermediate state

torantie committed:
intermediate state

torantie committed:
final update

torantie committed:
final update

torantie committed:
final update

torantie committed:
intermediate state.
todo
- individual load of meshes
- fix wireframe render
- add Mesh type text
- if possible fix mean value smoothing

torantie committed:
intermediate state.
Done
- individual load of meshes
- fix wireframe render
- fix normal calculation
- add Mesh type text

Todo
- if possible fix mean value smoothing

torantie committed:
intermediate state.
Done
- individual load of meshes
- fix wireframe render
- fix normal calculation
- add Mesh type text

Todo
- if possible fix mean value smoothing

torantie committed:
intermediate state.
Done
- main code cleanup
- changed io gui

Todo
- if possible fix mean value smoothing

torantie committed:
intermediate state.
Done
- halfedge.js code cleanup

Todo
- if possible fix mean value smoothing

torantie committed:
bug fix
- only loaded mesh for on one side
- didn't load quad mesh because count and not length was used

torantie committed:
bug fix:
- triangular mesh type wasn't detected properly

changed:
- default load object

added:
-demo objects

torantie committed:
changed bunny quad location

torantie committed:
Last commit adjusted one comment.
Some console logs were left in to check values of the mean value calculation and to show different tries for implementing mean value weight smoothing.

torantie committed:
ReadMe update
Forgot to upload deformed_plane_triangle.obj

changkun commented on projects/torantie/package.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "gp-3-smooth",
```
Remember to change this to your project name

changkun commented on projects/torantie/package.json:
```diff
@@ -0,0 +1,30 @@
+    "start": "webpack-dev-server --host localhost --open",
+    "build": "webpack --env.prod"
+  },
+  "author": "Changkun Ou",
```
Change the author to your name or your github username

changkun commented on projects/torantie/package.json:
```diff
@@ -0,0 +1,30 @@
+    "build": "webpack --env.prod"
+  },
+  "author": "Changkun Ou",
+  "license": "GPL-3.0",
```
Maybe more information about GPLv3 that might be good to know for your future open source work:

```
GNU GPLv3:
1. Anyone can copy, modify and distribute this software.
2. You have to include the license and copyright notice with each and every distribution.
3. You can use this software privately.
4. You can use this software for commercial purposes.
5. If you dare build your business solely from this code, you risk open-sourcing the whole code base.
6. If you modify it, you have to indicate changes made to the code.
7. Any modifications of this code base MUST be distributed with the same license, GPLv3.
8. This software is provided without warranty.
9. The software author or license can not be held liable for any damages inflicted by the software.
```

Full: https://github.com/mimuc/gp-ws2021/blob/master/LICENSE

changkun commented on projects/torantie/src/halfedge.js:
```diff
@@ -0,0 +1,881 @@
+// API Usage about @penrose/linear-algebra:
```
```
/**
 * Copyright © 2021 YOUR NAME <YOUR DOMAIN>. All rights reserved.
 * Portions Copyright © 2020-2021 Changkun Ou (contact@changkun.de)
 * 
 * Use of this source code is governed by a GNU GLPv3 license that can be
 * found in the LICENSE file.
 */

```

Add the above comments to the top of every source code file that you care about your originality.
Change the following information that fits your case:

1. Add your name and your domain (if possible)
2. Remove the `Portions ...` line if a file is not derived from my solution and completely newly added

changkun reviewed (changes requested):
Code haven't review yet, let's first get the licensing of your code right :)

torantie committed:
Added:
- package.json: project name, description, author name
- added copyright notice to all files that were edited for this project (main.js, renderer.js and halfedge.js)

changkun commented on projects/torantie/src/halfedge.js:
```diff
@@ -0,0 +1,895 @@
+// in node_modules/@penrose/linear-algebra/docs/*.html, but the above
+// information are all the APIs you need for this project.
+
+/**
```
Should be on top of the file starting from line 1 :)

torantie committed:
Changed:
- copyright notice moved to the top of the file

torantie committed:
Changed:
- project does not compile if project name has spaces. Changed project name.

changkun reviewed (approved):
- Nice teamwork on the quad support.
- Nice huge variety of comparison on different smoothing methods.

Regarding the mean value weights does not work, I just spend some time on reading an existing UnrealEngine 4.6's implementation: 

https://github.com/EpicGames/UnrealEngine/blob/5cf13708ed609c390e1cb2d27e2dd48c11423acd/Engine/Plugins/Experimental/GeometryProcessing/Source/DynamicMesh/Private/Solvers/PrecomputedMeshWeightData.cpp#L169-L208

https://github.com/EpicGames/UnrealEngine/blob/2bf1a5b83a7076a0fd275887b373f8ec9e99d431/Engine/Plugins/Experimental/GeometryProcessing/Source/DynamicMesh/Public/Solvers/PrecomputedMeshWeightData.h#L131-L142

https://github.com/EpicGames/UnrealEngine/blob/2bf1a5b83a7076a0fd275887b373f8ec9e99d431/Engine/Plugins/Experimental/GeometryProcessing/Source/DynamicMesh/Public/Solvers/LaplacianMatrixAssembly.h#L390-L455

Especially the thrid link. If you are interested in how they did it, it might be useful for you to read :)

From what I read at the moment, they did a lot precaution for numerical calculation, and there is an interesting prevention on boundary edges:

https://github.com/EpicGames/UnrealEngine/blob/2bf1a5b83a7076a0fd275887b373f8ec9e99d431/Engine/Plugins/Experimental/GeometryProcessing/Source/DynamicMesh/Public/Solvers/LaplacianMatrixAssembly.h#L437

Thus, I think overall your implementation is almost right, the weighted sum for the (ij, i != j) interior laplacian matrix, and the sum of all weights for the (ij, i==j) boundary laplacian.

By the way, I did a small typo change to your implementation:

```diff
-      weightTriplet.addEntry(sum/*/count*/, i, i)
+      weightTriplet.addEntry(-sum/*/count*/, i, i)
```

and I got a very interesting and promising result:

![image](https://user-images.githubusercontent.com/5498964/110215135-82dc1900-7ea8-11eb-9873-56d3c62f64eb.png)

![image](https://user-images.githubusercontent.com/5498964/110215190-bb7bf280-7ea8-11eb-8a9a-01d650f8a8db.png)

The overall shape is very close (especially the bunny in the second figure), but clearly there should (might) be some numerical issue that exists that prevents you from getting a good result.

torantie commented:
interesting change negating the sum in the weight triplet. The shape does seem more promising. What i kind of find weird is that with this change both the sum and the individual values (in the code called "w") are now negated.
```
vert.forEachHalfEdge(h => {
        let w = 0

        count++
        switch (weightType) {
          case 'uniform':
            w = 1
            break;
          case 'cotan':
            w = (h.cotan() + h.twin.cotan())/2
            break;
          case 'mean value':
            w = h.meanValueWeight()
            break;
        }


        sum += w
        weightTriplet.addEntry(-w, i, h.twin.vertex.idx)
      })
      weightTriplet.addEntry(sum/*/count*/, i, i)
```
Based on the Laplacian definition in the lecture only the sum should be negated.
![image](https://user-images.githubusercontent.com/14915789/110337258-0111fa00-8026-11eb-8d19-cbce7b6e42a9.png)
But since I (and the solution of exercise 5) used formula with Mass Matrix + Laplacian matrix, we negated the w not the sum.
`
let f = M.plus(W.timesReal(timeStep).timesReal(lambda))
`
If we use Mass Matrix minus Weight Martix we can negate the sum and leave the "w" without a minus.
At least these two versions where the only ones where uniform and cotan weights still worked. With the version where the w and sum are negated we get something like this for cotan for example, which seems kind of similar to the result of mean value weights:
![image](https://user-images.githubusercontent.com/14915789/110338385-39fe9e80-8027-11eb-818f-9e654f3abc8a.png)
Is also tested the version (both w and sum negated) with lower values of lambda, but it seems that the deformation is also very high when the weight matrix is scaled down. So I am not so sure if the change actually provides a better result.
![image](https://user-images.githubusercontent.com/14915789/110338207-0ae82d00-8027-11eb-822b-bc119afcacea.png)

Thank you for the unreal engine links. Unfortunately i am not able to access them (404 Page not found). Did the repository change or do i need something else to view their github?

changkun commented:
> Did the repository change or do i need something else to view their github?

The repo is private, you need to signup for the unreal engine to get access to the repository (free), there are some instructions:

https://www.unrealengine.com/en-US/ue4-on-github
